### ISSUE TITLE
Add Style/FeatureEnvy

### ISSUE THREAD

**Issue by bkuhlmann** [rubocop/rubocop#15118]
### Before submitting

- [x] I have searched the existing issues and this hasn't been requested yet.

### Is your feature request related to a problem?

I'd like to suggest adding [Reek](https://github.com/troessner/reek) code smells by starting with [FeatureEnvy](https://github.com/troessner/reek/blob/master/docs/Feature-Envy.md). Maybe this should be a new category like "Smell" instead of "Style"? Regardless -- and if we only focus on FeatureEnvy -- this would be a good start.

Accounting for code smells, like feature envy, would eventually eliminate the need to use Reek since it's hardly maintained these days but the code smell detection Reek provides is highly valuable.

### Describe the solution you'd like

I'd like the Style/FeatureEnvy cop to raise an error on code smells where an object in a method is referenced *four times* or more because that means the functionality should be moved to the object in question.

Note that, default Reek behavior, is to raise a feature envy code smell when an object is only referenced twice. In practice, this is too strict because it's quite common to only need to message an object twice for trivial things but once you start referencing the same object four or more times then, yes, you definitely have a code smell.

The default configuration could look like this:

``` yaml
Styles/FeatureEnvy:
  Enabled: true
  AllowedPatterns: []
  AllowedMethods: []
  Min: 3
```

### Code examples

```ruby
# No
def render_errors result, response
  response[:message] = "Demo"
  response.format = "text/html"
  response.flash[:alert] = result.errors
  response.redirect_to routes.path(:gallery)
end

# Yes (tolerable)
def render_errors result, response
  response[:message] = "Demo"
  response.flash[:alert] = result.errors
  response.redirect_to routes.path(:gallery)
end
```

### Describe alternatives you've considered

Nothing like this exists in RuboCop as of yet.

### Additional context

The reason the "No" example (above) is bad is because you are messaging the `response` object more than three times. This is a code smell because this behavior should be encapsulated within the `response` object. For example, the fix is to make the `response` object more elegant via a better Object API. Here's what the refactoring would look like after the code smell has been detected.

``` ruby
response.with(message: "Demo, format: "text/html", alert: result.errors)
        .redirect_to routes.path(:gallery)
```

**Comment by hammadxcm** [comment]
As someone interested in picking this up, a few questions before writing code, since the answers shape the design significantly:

1. **Department placement** — You raised this yourself. Is the maintainers' preference `Style`, `Metrics` (since this is fundamentally a counting heuristic akin to `Metrics/AbcSize`), or a new `Smell` department? A new department is a meaningful project-direction change and probably warrants its own discussion.

2. **Default enablement** — Should this be `Enabled: true` by default, or `pending`/`false`? Feature envy detection is inherently heuristic and opinionated; enabling by default would likely surface offenses in a large fraction of real codebases.

3. **Receiver-counting semantics** — The proposal counts references to an object, but there's a lot of nuance:
   - Should method chains (`response.foo.bar`) count as one reference to `response` or propagate?
   - Do `self` references count?
   - Are block-local receivers (`response.tap { |r| r.foo; r.bar }`) exempt?
   - How should the cop treat delegators, `Struct` members, and DSL-heavy code (Rails controllers, RSpec, builders) where repeated receiver usage is idiomatic rather than smelly?

4. **Overlap with existing cops** — Is there concern about overlap with `Metrics/AbcSize`, `Metrics/MethodLength`, or the Law-of-Demeter-adjacent style guidance?

Happy to take a stab at an implementation once there's a steer on 1 and 2 in particular.

**Comment by bkuhlmann** [comment]
Yeah, a few notes:

1. My preference would be `Smell` but is the maintainers call.
2. Yeah, this would have to be `false` by default.
3. Speaking, roughly, how Reek handles this, you'd need to:
   - Allow messages chains as they are a single reference.
   - Avoid `self` when used more than the minimum.
   - Avoid blocks that exceed the minimum.
   - Avoid delegators, Struct/Data members, and DSLs that exceed the minimum.
4. I don't think there is overlap because this would also encourage less lines of code. Maybe I'm missing something?

**Comment by hammadxcm** [comment]
Thanks @bkuhlmann — the receiver-counting semantics you outlined are useful and I think workable.

The blocking question for an implementation is still the department placement, which (as you note) is a maintainers' call rather than ours.

@koic @bbatsov — could one of you weigh in on whether this fits the project's direction? Specifically:

1. Are you open to RuboCop core absorbing Reek-style code-smell checks, or would these belong in an extension gem?
2. If absorbed, would you want a new `Smell` department (per the OP's preference), or is `Metrics` (most counting-based heuristic) or `Style` more appropriate?

Happy to pick up the implementation once there's a steer on the department. The cop body itself is non-trivial (method-chain coalescing, `self`/block-local exemption, delegator/Struct/DSL handling), so I'd rather not invest in a draft under the wrong namespace.

**Comment by paul-mothapo** [comment]
> Thanks [@bkuhlmann](https://github.com/bkuhlmann) — the receiver-counting semantics you outlined are useful and I think workable.
> 
> The blocking question for an implementation is still the department placement, which (as you note) is a maintainers' call rather than ours.
> 
> [@koic](https://github.com/koic) [@bbatsov](https://github.com/bbatsov) — could one of you weigh in on whether this fits the project's direction? Specifically:
> 
> 1. Are you open to RuboCop core absorbing Reek-style code-smell checks, or would these belong in an extension gem?
> 2. If absorbed, would you want a new `Smell` department (per the OP's preference), or is `Metrics` (most counting-based heuristic) or `Style` more appropriate?
> 
> Happy to pick up the implementation once there's a steer on the department. The cop body itself is non-trivial (method-chain coalescing, `self`/block-local exemption, delegator/Struct/DSL handling), so I'd rather not invest in a draft under the wrong namespace.

Co-Asking.

**Comment by andyw8** [comment]
I've started an experimental project to port Reek to be a RuboCop plugin. `FeatureEnvy` isn't yet covered, but perhaps you could open a PR.

https://github.com/andyw8/rubocop-reek
